[JSC] The fast path for JSONP should handle TDZ

hyjorc1 opened the pull request:
#### 303ef67fea1e5cc1a71d7510579ad2e41e07d533
<pre>
[JSC] The fast path for JSONP should handle TDZ
<a href="https://bugs.webkit.org/show_bug.cgi?id=243578">https://bugs.webkit.org/show_bug.cgi?id=243578</a>
rdar://95963298

Reviewed by Alexey Shvayka.

TDZ are misuses for const/let in JSONP. If we don&apos;t want to abandon the benefits of the JSONP
fast path for const/let, we should throw exceptions for handling TDZ.

* JSTests/stress/jsonp-program-evaluate-path-must-consider-global-lexical-environment.js:
* JSTests/stress/jsonp-program-evaluate-path-must-handle-tdz.js: Added.
(shouldThrow):
(test1):
(test2):
(test3):
(test4):
(test5):
(test6):
(test7):
(test8):
(test9):
(test10):
(test11):
(test12):
(test13):
(test14):
(test15):
(test16):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::executeProgram):

Canonical link: <a href="https://commits.webkit.org/253157@main">https://commits.webkit.org/253157@main</a>
</pre>

shvaikalesh reviewed (commented):
Could you please expand in commit message on why we want fast path to handle TDZ rather than tweaking JSONP parser not to take the fast path if there are `const` / `let`?

shvaikalesh reviewed (approved):
r=me

hyjorc1 committed:
[JSC] The fast path for JSONP should handle TDZ
https://bugs.webkit.org/show_bug.cgi?id=243578
rdar://95963298

Reviewed by Alexey Shvayka.

TDZ are misuses for const/let in JSONP. If we don't want to abandon the benefits of the JSONP
fast path for const/let, we should throw exceptions for handling TDZ.

* JSTests/stress/jsonp-program-evaluate-path-must-consider-global-lexical-environment.js:
* JSTests/stress/jsonp-program-evaluate-path-must-handle-tdz.js: Added.
(shouldThrow):
(test1):
(test2):
(test3):
(test4):
(test5):
(test6):
(test7):
(test8):
(test9):
(test10):
(test11):
(test12):
(test13):
(test14):
(test15):
(test16):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::executeProgram):

Canonical link: https://commits.webkit.org/253157@main
